robottelo/report_portal/portal.py

```diff
@@ -0,0 +1,222 @@
+        if launch_type not in launch_types:
+            raise ValueError(
+                f'Wrong Launch Type is given as {launch_type} should be one of {launch_types}'
+            )
+        params = {'page.page': 1, 'page.size': 500, 'page.sort': 'start_time'}
+        resp = requests.get(
+            url=f'{self.api_url}/launch', headers=self.headers, params=params, verify=False
+        )
+        resp.raise_for_status()
+        data = resp.json()
+        launches_ = {}
+        if data['content']:
+            all_launches = sorted(
+                data['content'], key=lambda launch: launch['start_time'], reverse=True
+            )
+            typed_launches = list(
+                filter(lambda launch: launch['name'].lower() == launch_type, all_launches)
+            )
+            for launch_info in typed_launches[:latest]:
+                version_compiler = re.compile(r'([\d\.]+)[\.-](\d+\.\d|[A-Z]+)')
+                launch_name = list(filter(version_compiler.search, launch_info['tags']))[0]
+                sat_ver = re.search(version_compiler, launch_name).group(1)
+                snap_version = re.search(version_compiler, launch_name).group(2)
+                if sat_ver in launches_:
+                    launches_[sat_ver].update({snap_version: launch_info})
+                else:
+                    launches_[sat_ver] = {snap_version: launch_info}
+        if sat_version:
+            if sat_version not in launches_:
+                raise ValueError(
+                    f'The given satellite version \'{sat_version}\' is not available in '
+                    f'Report portal project \'{self.rp_project}\''
+                )
+            else:
+                return launches_[sat_version]
+        return launches_
+
+    def launch(self, sat_version, snap_version=None, launch_type='satellite6'):
+        """Returns a specific launch data in Report Portal Project
+
+        This does not includes each tests data in launch
+
+        :param str sat_version: The satellite version
+        :param str snap_version: The snap version of a given satellite version
+            if None, the latest launch data of a given sat_version is returned
+        :param str launch_type: Either satellite6 or upgrades, default returns only
+            non-upgrade launches
+        :returns dict: The data directory of requested or latest launch
+        """
+        launches_ = self.launches(sat_version=sat_version, launch_type=launch_type)
+        if snap_version:
+            if snap_version not in launches_.keys():
+                raise ValueError(
+                    f'The given snap_version \'{snap_version}\' is not available '
+                    f'in satellite version \'{sat_version}\''
+                )
+        else:
+            # Get the latest launch of the given sat_version
+            snap_version = list(launches_.keys())[0]
+        return launches_[snap_version]
+
+
+class Launch:
```

Comment:
What's your intent for the use of this class?

I would think that the `ReportPortal.launches|launch` methods would be creating instances of this class, not just returning raw data structures.

Reply:
Not specific, but from your last comment `Launch.report_portal should be an instance attribute passed to its constructor from the ReportPortal instance it belongs to and as per our discussion to remove only Test class, I kept this. :)

I really don't think we should go into returning objects now, it's not our necessity right now. But having that will make it smarter.

What would you suggest ?

Reply:
@mshriver Also if we return Launch objects for launches and launch functions in ReportPortal, we will end up with maximum depth recursssion error and unless we create a separate functions that will return objects of launch. I also beleive that this is significantly add up to the time consumption.

Reply:
The recursion is only because you're trying to set a launch property on a launch class to the result of the function that should be creating the instance in the first place.

This `Launch` class should model a specific launch instance from Report Portal. It should not have a `launch` property on it, it IS a launch. This is seemingly trying to be a class that's an interface for a launch, but ReportPortal instance is our interface.

In `ReportPortal.launches` you have all of the info that should define a launch instance in report portal, and you return it as a dictionary within a list.  That should instead be taking all of that launch info, and returning a list of `Launch` instances.

A `Launch` should have attributes that uniquely define it, as given by the named `launch_info` variable in `ReportPortal.launches` - the name, sat version, and snap version.

The launch statistics could also be defined as a launch instance attribute, since they're available at init.

This is creating a needlessly layered pattern within your pytest fixture in #7758.  Its also assuming there's only one launch for a given sat version, which is not a safe assumption.

With what I'm suggesting, you should be able to reference the 'info' for a launch as instance attributes, and you shouldn't have to even directly instantiate a launch.

It would allow for the following pattern
```
rp = ReportPortal()
matching_launches = rp.Launches(sat_version=current_sat_version_under_test)
# maybe log that there were multiple launches, maybe composite them!
last_launch = matching_launches.pop()
# if statistics is an instance attribute, a dictionary:
last_pass_percent = round(100*(last_launch.statistics['passed']/last_launch.statistics['total']))
last_failed_tests = last_launch.tests(defect_type='to_investigate')
```

Reply:
ahh, Now I got it right! Or May be just back from summit to my regular work :)

Implemented !